File: frontend/src/lib/dmxPartyChannelGroups.ts
import type {DMXPartyChannelGroup, DMXPartyConfig} from "@/types/controller.ts";

export const PARTY_CHANNEL_GROUPS: ReadonlyArray<{
    id: DMXPartyChannelGroup;
    label: string;
    description: string;
}> = [
    {id: "movement", label: "Movement", description: "Pan, tilt, motor speed"},
    {id: "color", label: "Color", description: "Color wheel and RGB mix"},
    {id: "gobo", label: "Gobo", description: "Gobo wheel, indexing, rotation"},
    {id: "beam", label: "Beam", description: "Dimmer, shutter, zoom, focus, iris, frost"},
    {id: "effects", label: "Effects", description: "Prism and prism rotation"},
];

export function partyChannelGroupEnabled(
    config: DMXPartyConfig,
    group: DMXPartyChannelGroup,
): boolean {
    const groups = config.channelGroups;
    if (!groups) {
        return true;
    }
    const value = groups[group];
    return value !== false;
}

export function togglePartyChannelGroup(
    config: DMXPartyConfig,
    group: DMXPartyChannelGroup,
    enabled: boolean,
): Partial<DMXPartyConfig> {
    return {
        channelGroups: {
            ...config.channelGroups,
            [group]: enabled,
        },
    };
}

File: internal/controller/dmx_party_channel_groups.go
package controller

import "strings"

const (
	partyChannelGroupMovement = "movement"
	partyChannelGroupColor    = "color"
	partyChannelGroupGobo     = "gobo"
	partyChannelGroupBeam     = "beam"
	partyChannelGroupEffects  = "effects"
)

func partyChannelGroupForType(normType string) string {
	switch normType {
	case "pan", "panfine", "tilt", "tiltfine", "infinitepan", "infinitetilt", "movementspeed":
		return partyChannelGroupMovement
	case "colorwheel", "colorcomponent", "colortemperature", "greensaturation", "xfadetocolor":
		return partyChannelGroupColor
	case "gobowheel", "goboindexing", "goboindexingfine", "goborotation", "goborotationfine", "goboshake":
		return partyChannelGroupGobo
	case "dimmer", "dimmerfine", "shutterstrobe", "zoom", "zoomfine", "focus", "focusfine",
		"iris", "irisfine", "frost", "frostfine", "onoff", "lamp":
		return partyChannelGroupBeam
	case "prism", "prismindexing", "prismindexingfine", "prismrotation":
		return partyChannelGroupEffects
	default:
		return ""
	}
}

func partyChannelGroupEnabled(cfg DMXPartyConfig, group string) bool {
	if group == "" {
		return true
	}
	if cfg.ChannelGroups == nil {
		return true
	}
	if v, ok := cfg.ChannelGroups[group]; ok {
		return v
	}
	return true
}

func partyChannelGroupAllowed(cfg DMXPartyConfig, normType string) bool {
	group := partyChannelGroupForType(strings.ToLower(strings.TrimSpace(normType)))
	return partyChannelGroupEnabled(cfg, group)
}
